### ISSUE TITLE
T37: Complete and adjudicate the visual development experiment

### ISSUE THREAD

**Issue by Sino-Huang** [Sino-Huang/multimodality_on_planning#75]
**Parent:** #38

## Completed execution scope

The operator continued the already-running cost-ranked v5 experiment instead of switching to the subsequently proposed four-hour pilot. This ticket records completion and adjudication of that actual run. The 512-record/one-epoch pilot was not executed and is not claimed complete. Original comments and frozen receipts remain preserved.

Contract: `issue-75-visual-development-32k-v5`, run `outputs/visual_development/issue75-32k-v5/attempt-001`.

## Result

All four training runs and all 864 selected episodes completed in 76.23 hours. The retained outcome is **VALID_STOP**, `scientific_completion: false`, because BFWS learned success was 66.7%, below the frozen 80% threshold. This is complete experimental evidence with a failed performance gate. Closing the execution ticket does not turn that result into PASS or lower a threshold.

## Executed protocol

- 43,876 source training records across BFS, BFWS, additive w3 and additive greedy; two epochs; one training seed, 17; final adapters only.
- Cost-selected 42 development task groups / 54 algorithm-task cases, preserved across conditions.
- Five evaluation seeds (17, 29, 43, 71, 101); exact reference once per case. Counts: 54 exact-reference, 270 random-valid, 270 pretrained-base and 270 SFT episodes.
- Complete state/goal observations, bounded Search Memory, operation budgets and independent trace replay. Source corpus, original splits and freezes remain unchanged.
- Two GPUs with distinct MASTER_PORT values 18575 and 18576. Actual elapsed time is retained; no four-hour budget compliance is claimed.

## Completion criteria

- Verify all four final adapter files and complete task/algorithm/condition/seed membership.
- Independently replay all 864 episodes and reconcile the saved metrics and gate verdict without new model calls or rewriting original outputs.
- Report success, invalid operations, budget usage, learned-versus-control gains and uncertainty limitations, including negative findings and control saturation.
- Report actual elapsed time and available resource evidence honestly. Whole-run v5 peak memory was not saved; reused qualification memory is calibration evidence only.
- Preserve the failed scientific gate and do not rerun merely to obtain PASS. Distinguish completed execution from performance-gate success and from the unrun deadline pilot.
- Hand off the actual training scope to #76/#77 before matched-modality claims; do not automatically enlarge downstream compute.

## Findings and handoff

SFT success: BFS 100%, BFWS 66.7%, additive w3 100%, additive greedy 100%. Random-valid success: 70.7%, 52%, 100%, 100%, respectively. Exact references pass all cases; the base condition has 0% success and 100% invalid operations under the frozen protocol. Only BFS has a positive saved whole-instance bootstrap lower bound against its best control. Additive controls saturate. Five rollout seeds do not establish training-seed variance.

The planned 512-record/one-epoch #76 pilot is not training-matched to these v5 adapters. Reconcile scope before #76/#77/#93–#95 comparison claims, preserving existing future spending caps. No multimodal, DAgger, final-test, replication or transfer completion is implied.

Evidence: `docs/experiments/issue75/v5-completion.md` and `v5-completion-verification.json`. Original run evidence remains under the run directory above.

## Blocked by
#73 (completed prerequisite)

**Comment by Sino-Huang** [comment]
Apply the resource-bounded evaluation protocol established by #54:

- drive incremental episode sessions in deterministic rounds and use batched Transformers generation with one shared backbone per GPU, adapter switching, adapter-isolated caching, and at most one request per active episode per round;
- freeze a 2x matching-reference decision-call limit separately from expansion limits; exhaustion is a valid unsuccessful episode;
- evaluate only final checkpoints across five seeds, with base, learned, random-valid, and exact-reference conditions on the identical selected tasks;
- qualify throughput without success outcomes before rollout, using the actual precision/batch settings and a cost-qualified fallback panel if full coverage cannot fit;
- stop before new launches at the wall-clock cutoff, preserve completed episodes, replay every completed episode, and never let partial coverage satisfy the gate.

Report invariant-valid success, invalid-operation rate, budget usage, and learned-vs-best-control gain separately. #54 reached 100% learned success on its reduced panel but still correctly `VALID_STOP` because random-valid also reached 100%; avoid interpreting panel success alone as structural advantage.

**Comment by Sino-Huang** [comment]
Supervisor time-budget clarification: this ticket must not require multiple independent training-seed replicates. For each distinct model/cell configuration, authorize exactly one training run using seed 17 and evaluate only the final checkpoint from that run.

Any existing five-seeds requirement must be interpreted as rollout/reference/evaluation seeds applied to the single trained checkpoint where repeated evaluation is still useful, not as five separately trained checkpoints. Do not estimate or claim training-seed variance, and do not substitute rollout-seed variation for it; use whole-problem-instance uncertainty where applicable.

If a frozen scientific criterion truly requires independent training-seed replication, issue the appropriate governed VALID_STOP or gated-scope receipt instead of launching additional training runs.

**Comment by Sino-Huang** [comment]
Verified the operator's completed `issue-75-visual-development-32k-v1/attempt-001` against its frozen configuration, authorization, gate binding, worker logs, and final receipt.

- Outcome: **VALID_STOP**, `gated-not-run`, `scientific_completion: false`.
- Elapsed: **3,609.47 seconds** (about 60 minutes). Qualification exhausted its frozen 3,600-second allowance.
- Both GPU workers last reported **4 / 31** completed qualification probes. Worker 0 emitted the stop receipt; the parent stopped the sibling worker.
- Authorization and gate bindings match the attempt.
- No complete qualification report, references, training, evaluation, or adjudication artifacts exist. The visual development matrix did not reach its scientific execution stages.

**Keep #75 open.** Its requirement for all declared visual conditions and references with full episode evidence is not satisfied. This is a governed resource stop, not evidence of model competence or failure of a learned policy.

The immutable attempt is retained at `outputs/visual_development/issue75-32k-v1/attempt-001/result.json`. Local verification is recorded in `docs/experiments/issue75/attempt-001-verification.json`. Do not use `--resume` to reset this completed attempt. The next step is to address qualification runtime and prepare a separately authorized successor, retaining the scientific coverage requirements and original evidence.

**Comment by Sino-Huang** [comment]
Investigated and fixed the qualification timeout locally in commit `0d24e30`.

The operator explicitly requested removing the authorization workflow for this research experiment. New #75 runs therefore require no approval/authorization file or receipt. They retain configuration snapshots, complete-coverage checks, runtime validity checks and ordinary resource limits. Historical receipts and the stopped v1 attempt remain preserved.

Diagnosis: v1 scheduled 940 generation calls per GPU by qualifying all three modalities and repeating scalar generation for duplicate entries in mixed batches. The revised visual-only qualification schedules 151 calls (83.9% fewer), retaining all 31 selected probe records, every-position scalar/batch/repeated semantic checks, 384-token timing and the training hardware probe. A fast simulated-clock regression reproduced the old cutoff failure and passes after the fix. Source scanning took 5.02 seconds; the first four CPU teacher replays took 1.86 seconds.

The separate one-hour qualification cutoff is removed; qualification uses the overall run deadline. Each generation operation is logged, completed probe timings are saved immediately, and 20-second heartbeats retain current progress. `--output` selects a fresh run directory without an approval document.

Validation: 24 focused tests, Black, Ruff, BasedPyright and the complete command dry-run pass. No new long GPU run was launched, so revised wall-clock throughput and scientific completion remain unmeasured. #75 stays open.

The default command now targets `outputs/visual_development/issue75-32k-v2/attempt-001`:

```bash
source ~/cd_vlaplan
python -u scripts/run_visual_issue75.py all
```

**Comment by Sino-Huang** [comment]
Fixed the v2 CUDA OOM locally in `ef6e41b`, without stopping or modifying process 49552.

Two causes were reproduced in bounded live GPU checks:

1. The failing record at batch eight OOMed in a fresh worker on GPU 0. Reducing that batch to two completed 384 output tokens with 41.4 GiB peak PyTorch allocation.
2. The largest selected visual probe (16,795 input tokens) also OOMed at batch one: native float32 GQA selected math attention and requested a 33.63 GiB tensor. Explicit KV-head expansion plus PyTorch efficient SDPA avoids this quadratic allocation while retaining float32 and Transformers' causal/padding masks. The same probe then completed 384 output tokens at 41.9 GiB peak allocation on GPU 0.

The new defaults use batch two / 24,000 padded batch tokens, qualify largest inputs first, and retain per-probe memory measurements. Tasks, observations and training settings remain unchanged. A live mixed-length scalar/batch/repeated semantic comparison passed on GPU 1 at 40.7 GiB peak allocation. Process 49552 continued using its existing GPU memory throughout.

30 focused tests, formatting/lint/type checks and the command dry-run pass. Measurements and explicit GPU/MASTER_PORT mappings are recorded in `docs/experiments/issue75/memory-fix.json`. These are bounded memory/semantic checks, not a completed matrix; keep #75 open.

The next run uses a fresh `outputs/visual_development/issue75-32k-v3/attempt-001` directory:

```bash
source ~/cd_vlaplan
python -u scripts/run_visual_issue75.py all
```

**Comment by Sino-Huang** [comment]
Investigated the v3 `neither frozen panel fits the clock` stop. Both GPU qualifications passed all 31 probes and the training hardware probes. The stop came from the admission estimate: it applied the maximum measured forced-384-token call cost (93.0 seconds) and training microstep cost (12.0 seconds) across the whole workload, assumed maximum episode call usage, then rejected both panels against a hard 20-hour budget.

These assumptions produced stress projections of approximately 431 days for full coverage and 76 days for fallback coverage. They are not measured ETAs; some overheads are also excluded, so they are not guaranteed wall-time bounds. There is no evidence here that the full matrix fits within 20 hours.

The research runner now defaults to `budget_mode: advisory`: preserve the full 97-task-group dev panel, display both stress projections, and do not use an estimate or elapsed-time cap to prevent execution. `budget_mode: hard` retains the existing time limits and full/fallback admission policy. Per-episode budgets, hardware/semantic checks and full-coverage requirements remain enforced.

V4 reuses the completed v3 hardware reports when model, corpus, training and device settings match, preserving the source contract IDs. This avoids repeating the 76-minute qualification run. Missing/failed hardware reports or changed execution settings cannot be reused.

Validation: 37 focused tests, formatting/lint/type checks, the full CLI dry-run, and a bounded public qualification-reuse check without child/GPU launches passed. Evidence: `docs/experiments/issue75/budget-fix.json`.

This fixes the pre-run budget rejection; it does not make the workload faster or establish scientific completion. No long run was launched. Keep #75 open pending execution and adjudication.

The default command now targets `outputs/visual_development/issue75-32k-v4/attempt-001`:

```bash
source ~/cd_vlaplan
python -u scripts/run_visual_issue75.py all
```

**Comment by Sino-Huang** [comment]
Implemented the requested per-instance cost pruning and pushed it to `main` in `8d2d170`.

The shared panel retains **42 of 97 dev task groups**, excludes **55**, and reduces the input-weighted evaluation work proxy by **85.0%**. It contains **54 algorithm/task cases / 864 condition episodes**: 15 BFS, 15 BFWS, 12 additive w3 and 12 additive greedy cases. Every available domain/algorithm-family combination remains represented.

Costs use each collected reference decision's complete visual input size with a per-algorithm curve from retained GPU timings, rather than applying the global slowest call to every instance. Selection uses computational cost only. Shared semantic problem groups across algorithms are kept or removed together, including all additive pairs. The same saved task IDs apply to text, visual and multimodal comparisons.

Manifest: `configs/experiments/issue75/cost-panel-v1.json`. It records all selected/excluded candidates, ranks, costs, reasons, reference bindings, input statistics, calibration sources and matching groups. The runner enforces it for references and rollout, uses the same cost-based GPU task placement as the estimate, and snapshots it for resume. The new default run directory is `outputs/visual_development/issue75-32k-v5/attempt-001`; passed v3 hardware qualification is reused.

Training is accounted for separately: the 43,876 training records and full-dev teacher-loss diagnostics remain unchanged. The conservative evaluation proxy drops from about 129.4 to 19.4 days; the training/diagnostic maximum-microstep stress proxy is about 11.25 days. These are not measured ETAs or an end-to-end guarantee. The coverage-preserving base still exceeds the 15-hour evaluation target, explicitly reported as `fits_evaluation_budget: false`; no domain is silently dropped to force a fit.

Validation: **44 focused tests**, Black, Ruff, BasedPyright, both CLI dry-runs and qualified panel binding passed. Tests cover identical modality membership, shared algorithm groups, outcome-blind selection, source coverage, runtime filtering and resume snapshots. No long GPU run was started. #75 remains open pending the actual selected matrix and adjudication; no #76 execution is claimed.

```bash
source ~/cd_vlaplan
python scripts/prepare_visual_cost_panel.py --dry-run
python -u scripts/run_visual_issue75.py all
```

**Comment by Sino-Huang** [comment]
Verified the operator’s completed v5 experiment and closing #75 as **completed execution with a failed scientific performance gate**.

- Four final adapters retained; 43,876 training records across four algorithms, two epochs, training seed 17.
- Full selected coverage: **42 task groups, 54 algorithm/task cases, 864 episodes** (54 exact, 270 random-valid, 270 base, 270 SFT).
- Independently replayed all 864 retained traces on CPU and matched every field of the saved adjudication; no new model calls or rewritten attempt outputs.
- Passed saved hardware qualification and all 31 trained-adapter probes on each GPU; final checkpoint and condition/seed coverage checks pass.
- Actual run time: **76.23 hours**. This is the older cost-ranked v5 experiment; the four-hour pilot was not run.

| Algorithm | SFT success | Random-valid success |
| --- | ---: | ---: |
| BFS | 100% | 70.7% |
| BFWS | 66.7% | 52.0% |
| Additive w3 | 100% | 100% |
| Additive greedy | 100% | 100% |

Exact references succeed on all cases; base success is 0% with 100% invalid operations under the frozen protocol. Only BFS has a positive saved whole-instance bootstrap lower bound against its best control. Additive controls saturate; no learned advantage is established there.

**The original result remains VALID_STOP / scientific_completion: false.** BFWS failed the frozen 80% learned-success threshold. Complete data collection does not make that scientific gate pass. No repeat-until-positive run is required. Whole-run v5 peak memory was not recorded; reused qualification memory is calibration evidence only.

The issue body now accurately describes the executed v5 scope. #76/#77 have the handoff: reconcile the full-corpus/two-epoch visual training with their planned 512-record/one-epoch pilot before claiming a matched modality comparison. Existing future spending caps remain; no downstream completion or larger run is implied.

Evidence: [completion report](https://github.com/Sino-Huang/multimodality_on_planning/blob/main/docs/experiments/issue75/v5-completion.md), [independent verification](https://github.com/Sino-Huang/multimodality_on_planning/blob/main/docs/experiments/issue75/v5-completion-verification.json). Original receipts and traces remain at `outputs/visual_development/issue75-32k-v5/attempt-001/`.
